enhancements/20250203-third-party-storage-support.md

```diff
@@ -0,0 +1,98 @@
+
+### User Experience In Detail
+
+This enhancement will abstract the Virtual Machine Image path generically. So, the user only needs to choose the corresponding StorageClass provided by the third-party storage solution. They can create a VM Image with their storage solution. Then, they can create a VM with the VM Image.
+
+### API changes
+
+TBD
+
+## Design
+
+### High-level Architecture
+
+![3rd-party-high-level](./20250203-third-party-storage-support/high-level-design.png)
+
+### Implementation Overview
+
+We could split the implementation into the following parts:
+- Introduce Containerized Data Importer (CDI) to Harvester
+- VM Image Creation through CDI
+- Root Disk Creation through CDI
+
+#### Introduce Containerized Data Importer (CDI) to Harvester
+
+To introduce Containerized-Data Importer (CDI) to Harvester, we need to add the charts and the CRDs to Harvester. The CDI will be responsible for the VM Image and the root disk creation with the third-party storage solution. 
+
+#### VM Image Creation through CDI
+
+The VM Image creation design could be like the following:
+![vm-image-design](./20250203-third-party-storage-support/vm-image-design.png)
+
+Because the target PVC (golden image) size should be defined before DataVolume creation, we need to prefetch the size and virtual size.
+We have different mechanisms for different source types (download/upload) for size/virtual size prefetching, and progress updates.
+
+For the golden image, we need to add webhook to protect the golden image from being deleted. The golden image should be protected until the VM Image is deleted.
+
+Also, we should not display the golden image on the volume page in the UI.
+
+#### Root Disk Creation through CDI
+
+The VM root disk creation design could be like the following:
+![root-disk-design](./20250203-third-party-storage-support/vm-root-disk-design.png)
+
+The root disk creation is similar to the VM Image creation. Now, with the golden image, we use the golden image as the source (type PVC), and the storage (target) is the root disk configuration. The root disk will be created with the CDI efficient cloning.
+
+One thing we need to take care of is that the root disk created by the CDI should be deleted with the DataVolume. If we only delete the PVC, the CDI will recreate it again with the DataVolume configuration.
+
+### Test plan
+
+- Ensure the CDI is installed default in Harvester
+- Ensure the fleet did not complain anythin with the new CDI. (Both fresh install and upgrade)
+- Ensure the VM image can be created through download/upload with different third party storage solutions (e.g. LVM, Rook, etc.)
+- Ensure the VM can be create with the above VM image (means the root disk is created through the third party storage solution)
+
+### Upgrade strategy
+
+This enhancement will add extra fields to the current VirtualMachineImage CRD. So, the upgrade needs to handle these new fields.
```

Comment:
Can we add a sample of what the new VMI YAML will look like?

Reply:
Already updated in the API change.